src/server/routing/auth/index.js

```diff
@@ -3,71 +3,121 @@ import { join } from "node:path";
+}
+
+/**
+ * Set up all auth methods for this platform.
+ */
 export function addPassportAuth(app) {
   app.use(passport.initialize());
   app.use(passport.session());
+  addGoogleAuth(app);
   addGithubAuth(app);
   addEmailAuth(app);
+  app.use(`/auth/logout`, (req, res, next) =>
+    req.logout((err) => {
+      if (err) return next(err);
+      res.redirect(`/`);
+    }),
+  );
 }
 
 /**
- * Set up github auth
+ * ...docs go here...
+ * @param {*} req
+ * @param {*} accessToken
+ * @param {*} refreshToken
+ * @param {*} profile
+ * @param {*} done
+ * @returns
  */
-function addGithubAuth(app) {
-  const githubStrategy = new GitHubStrategy(
-    githubSettings,
-    (req, accessToken, refreshToken, profile, done) => {
-      const { username, slug } = req.session.reservedAccount ?? {};
-
-      const userObject = {
-        slug,
-        profileName: profile.displayName,
-        service: profile.provider,
-        service_id: profile.id,
-      };
-
-      let user;
+function processOAuthLogin(req, accessToken, refreshToken, profile, done) {
+  const { user: sessionUser } = req.session.passport ?? {};
+
+  // Are there special circumstances surrounding this callback?
+  const { username, slug, newProvider } = req.session.reservedAccount ?? {};
+  delete req.session.reservedAccount;
+  req.session.save();
+
+  const userObject = {
+    profileName: profile.displayName,
+    service: profile.provider,
+    service_id: profile.id,
+  };
+
+  let user;
+
+  // If we have a user slug, this is a new account signup
+  if (username && slug) {
+    user = processUserSignup(username, userObject);
+  }
+
+  // If we have a new provider name, we need to add this
+  // as an additional login provider for this user's account
+  else if (newProvider) {
+    user = addLoginProviderForUser(sessionUser, userObject);
+  }
+
+  // If not, this is a regular login, where we need to find
+  // the user that belongs to this service profile.
+  else {
+    user = processUserLogin(userObject);
+  }
+
+  return done(null, user);
+}
 
-      // If we have a user slug, this is a new account signup
-      if (userObject.slug) {
-        user = processUserSignup(username, userObject);
-      }
+/**
+ * Set up google auth
+ */
+function addGoogleAuth(app) {
+  const googleStrategy = new GoogleStrategy(googleSettings, processOAuthLogin);
 
-      // If not, this is a log-in, where we need to find
-      // the user that belongs to this service profile.
-      else {
-        user = processUserLogin(userObject);
-      }
+  passport.use(googleStrategy);
 
-      return done(null, user);
-    },
+  const google = Router();
+  google.get(`/error`, (req, res) => res.send(`Unknown Error`));
+  google.get(`/callback`, handleGoogleCallback, (req, res) =>
+    res.redirect(`/`),
   );
+  google.get(`/logout`, logout);
+  google.get(`/`, loginWithGoogle);
+  app.use(`/auth/google`, google);
+}
+
+/**
+ * Set up github auth
+ */
+function addGithubAuth(app) {
```

Comment:
We need to exit early if the settings object is undefined, here and in the other functions